### PR TITLE
Use atomic operations for mInitOK to fix a data race

### PR DESCRIPTION
**Pull request by Naios**
* Requires C++11
* WARNING: ThreadSanitizer: data race (pid=1023)
  Write of size 1 at 0x7b1400000470 by main thread (mutexes: write M90):
    -0 efsw::FileWatcherGeneric::~FileWatcherGeneric() efsw/src/efsw/FileWatcherGeneric.cpp:20:10
    -1 efsw::FileWatcherGeneric::~FileWatcherGeneric() efsw/src/efsw/FileWatcherGeneric.cpp:19:1
    -2 efsw::FileWatcher::~FileWatcher() efsw/src/efsw/FileWatcher.cpp:78:2
  Previous read of size 1 at 0x7b1400000470 by thread T5:
    -0 efsw::FileWatcherGeneric::run() efsw/src/efsw/FileWatcherGeneric.cpp:151:8
    -1 efsw::Private::ThreadMemberFunc<efsw::FileWatcherGeneric>::run() efsw/src/efsw/Thread.hpp:81:22
    -2 efsw::Thread::run() efsw/src/efsw/Thread.cpp:48:15 (libidle.so+0xf8abab)
    -3 efsw::Platform::ThreadImpl::entryPoint(void*)

The reason why thread sanitizer yields a warning about this is that you are using mInitOK as atomic flag that should be visible on all threads. While (to my understanding) a relaxed memory order gets translated to an ordinary move on x86, it probably does not make any difference to a non atomic variable, however it is required for correctness on all platforms.

I do not know which C++ standard you are currently requiring, therefore the AtomicBool class probably requires an implementation for posix and Windows if you are still supporting platforms below C++11.

### CODE REVIEW

**Comment by SpartanJ** [comment]
Hi @Naios !
Thanks for your contribution.
I'm not sure to merge it. The thing is that I never restricted the C++ version, I'm almost sure that efsw compiles with any C++98 compatible compiler (otherwise it will be any C++03). There was at least one company using efsw in very old setups (https://github.com/SpartanJ/efsw/issues/80). I kind of prefer not to break anyone's setup than adding a fix to a non-issue. The thread sanitizer is correct but in this case, I don't see how this could cause an issue. `mInitOK` is just used to keep the watcher thread alive, and the worst-case scenario would produce an extra iteration in the endless loop.
So, in conclusion: if you think this is an important issue I may merge it. But I would prefer to avoid breaking old C++ compatibility.
BTW: I would avoid using the generic watcher unless there's a very good reason to use it. It's resource-intensive and inefficient.

**Comment by Naios** [comment]
Thank you for your quick feedback. I understand that you still want to support C++98. Therefore I have added a fallback option that uses `volatile` now and the `std::atomic` is guarded behind the `EFSW_USE_CXX11` definition now. This should improve the issue mentioned above without breaking anyone's code.
Providing platform-specific fence instructions is probably out of scope because the list of supported platforms seems quite large to me.

I am using the generic file watcher only to bypass restrictions on VMs running inside docker through the WSL on Windows (which unfortunately does not support native file watchers),

**Comment by SpartanJ** [comment]
Can you also add this new flag in the premake files? If you don't know how I can add it later.

**Comment by Naios** [comment]
Could you add the definitions to premake later? I do not have worked with premake yet and cannot test it properly.

**Comment by SpartanJ** [comment]
No problem, thanks for your contribution!
